core/src/test/scala/quasar/plugin/postgres/datasource/PostgresDatasourceSpec.scala

```diff
@@ -194,44 +196,158 @@ object PostgresDatasourceSpec
+      } yield ws must containTheSameElementsAs(expected)
+    }
 
-        qr <- datasource.evaluate(InterpretedRead(path, ScalarStages.Id))
+    "noop when a masked path doesn't begin with a field" >>* {
+      val mask: ScalarStage = ScalarStage.Mask(Map(
+        CPath.parse("[2].serial") -> Set(ColumnType.String),
+        CPath.parse("[2].height") -> Set(ColumnType.Number)))
 
-        ws <- qr match {
-          case QueryResult.Typed(_, s, _) =>
-            s.chunks.parseJsonStream[Json]
-              .map(_.as[Widget].toOption)
-              .unNone
-              .compile.toList
+      val stages = ScalarStages(IdStatus.ExcludeId, List(mask))
 
-          case _ =>
-            IO.pure(List.empty[Widget])
-        }
-      } yield ws must containTheSameElementsAs(widgets)
+      for {
+        _ <- loadWidgets(maskWidgets)
+        r <- resultsOf[Widget](widgetsRead(stages))
+      } yield r must (stages, maskWidgets).zip(be_===, containTheSameElementsAs(_))
+    }
+
+    "noop when mask stage isn't first" >>* {
+      val mask: ScalarStage = ScalarStage.Mask(Map(
+        CPath.parse(".foo.serial") -> Set(ColumnType.String),
+        CPath.parse(".foo.height") -> Set(ColumnType.Number)))
+
+      val stages = ScalarStages(IdStatus.ExcludeId, List(ScalarStage.Wrap("foo"), mask))
+
+      for {
+        _ <- loadWidgets(maskWidgets)
+        r <- resultsOf[Widget](widgetsRead(stages))
+      } yield r must (stages, maskWidgets).zip(be_===, containTheSameElementsAs(_))
+    }
+
+    "eliminates mask stage when all paths are a single field and type is top" >>* {
+      val mask: ScalarStage = ScalarStage.Mask(Map(
+        CPath.parse(".width") -> ColumnType.Top,
+        CPath.parse(".height") -> ColumnType.Top))
+
+      val expected = List(
+        Json("width" := 1.1, "height" := 2.2),
+        Json("width" := 3.3, "height" := 4.4),
+        Json("width" := 5.5, "height" := 6.6),
+        Json("width" := 7.7, "height" := 8.8))
+
+      for {
+        _ <- loadWidgets(maskWidgets)
+        r <- resultsOf[Json](widgetsRead(ScalarStages(IdStatus.ExcludeId, List(mask))))
+      } yield r must (ScalarStages.Id, expected).zip(be_===, containTheSameElementsAs(_))
+    }
+
+    "retains mask stage when any path has length > 1" >>* {
+      val mask: ScalarStage = ScalarStage.Mask(Map(
+        CPath.parse(".width[2]") -> ColumnType.Top,
+        CPath.parse(".height") -> ColumnType.Top))
+
+      val stages = ScalarStages(IdStatus.ExcludeId, List(mask))
+
+      val expected = List(
+        Json("width" := 1.1, "height" := 2.2),
+        Json("width" := 3.3, "height" := 4.4),
+        Json("width" := 5.5, "height" := 6.6),
+        Json("width" := 7.7, "height" := 8.8))
+
+      for {
+        _ <- loadWidgets(maskWidgets)
+        r <- resultsOf[Json](widgetsRead(stages))
+      } yield r must (stages, expected).zip(be_===, containTheSameElementsAs(_))
+    }
+
+    "retains mask stage when any path has type != top" >>* {
```

Comment:
Checking my understanding here: in this example we verify that we successfully push down masking, and we verify that the mask stage is retained. Do we retain the mask stage because we don't want to rely on the datasource having necessarily masked, and it'll basically be a no-op when the data gets to tectonic?

Reply:
It is because the mask type is something more specific than `Top` so we're punting on that and allowing tectonic to deal with the type.

When used with REFORM, the type masking seems superfluous with this datasource since tables are homogeneous and REFORM adds typechecks based on the discovered types from the data browser (so they'll always match the column type).

So retaining these in our use-case isn't strictly necessary, but I was focused on doing as little as possible that still realizes the column masking benefit. We could check all the columns against the table metadata in order to eliminate the mask if desired or if there are performance gains to be had by tectonic not having to process the masking.

Reply:
Would it be worth having a test with
```scala
      val mask: ScalarStage = ScalarStage.Mask(Map(
        CPath.parse(".width") -> Set(ColumnType.String),
        CPath.parse(".height") -> ColumnType.Top))
```
with the same expected results. This would show that `".width"` data is retained even though it's of the wrong column type.

Reply:
Sure, I'll add it :+1:

Reply:
Cool, I think it'll be good documentation.

Reply:
To amend my earlier statement, for `json` and `jsonb` postgres columns the data could be heterogeneous between rows. So there are some cases where masking wouldn't be identity.